src/Build/Evaluation/IntrinsicFunctions.cs

```diff
@@ -629,10 +629,30 @@ internal static bool AreFeaturesEnabled(Version wave)
             return ChangeWaves.AreFeaturesEnabled(wave);
         }
 
-        internal static string SubstringByTextElements(string input, int start, int length)
+        internal static string SubstringByAsciiChars(string input, int start, int length)
         {
-            StringInfo stringInfo = new StringInfo(input);
-            return stringInfo.SubstringByTextElements(start, length);
+            if (start > input.Length)
+            {
+                return string.Empty;
+            }
+            if (start + length > input.Length)
+            {
+                length = input.Length - start;
+            }
+            StringBuilder sb = new StringBuilder();
+            for (int i = start; i < start + length; i++)
+            {
+                char c = input[i];
+                if (c >= 32 && c <= 126)
```

Comment:
With breaking multichar sequences, we can probably as well end up with chars that are not a valid path chars.
Can you please as well check here that the char is not an invalid file name char? You can check against https://github.com/dotnet/msbuild/blob/1ce195e7cb119cdf66e2680badc29fcc4b6e20df/src/Shared/FileUtilities.cs#L107